trin-execution/src/subcommands/era2.rs

```diff
@@ -0,0 +1,243 @@
+                storage_count: storage_count as u32,
+            }))?;
+
+            for storage_chunk in storage.chunks(MAX_STORAGE_ITEMS) {
+                era2.append_entry(&AccountOrStorageEntry::Storage(StorageEntry(
+                    storage_chunk.to_vec(),
+                )))?;
+            }
+
+            accounts_processed += 1;
+            if accounts_processed % 10000 == 0 {
+                info!("Processed {} accounts", accounts_processed);
+            }
+        }
+
+        info!("Era2 snapshot exported");
+
+        Ok(())
+    }
+}
+
+pub struct StateImporter {
+    pub trin_execution: TrinExecution,
+    importer_config: ImportStateConfig,
+}
+
+impl StateImporter {
+    pub fn new(trin_execution: TrinExecution, importer_config: ImportStateConfig) -> Self {
+        Self {
+            trin_execution,
+            importer_config,
+        }
+    }
+
+    pub fn import_state(&mut self) -> anyhow::Result<()> {
+        info!("Importing state from .era2 file");
+        if self.trin_execution.next_block_number() != 0 {
+            return Err(Error::msg(
+                "Cannot import state from .era2, database is not empty",
+            ));
+        }
+
+        let mut era2 = Era2::open(self.importer_config.path_to_era2.clone())?;
+        info!("Era2 reader initiated");
+        let mut accounts_imported = 0;
+        while let Some(account) = era2.next() {
+            let AccountOrStorageEntry::Account(account) = account else {
+                return Err(Error::msg("Expected account, got storage entry"));
+            };
+            let AccountEntry {
+                address_hash,
+                account_state,
+                bytecode,
+                storage_count,
+            } = account;
+
+            let account_state: Account = account_state.into();
+
+            // Build storage trie
+            let account_db = AccountDB::new(address_hash, self.trin_execution.database.db.clone());
+            let mut account_trie = EthTrie::new(Arc::new(account_db));
+            for _ in 0..storage_count {
+                let Some(AccountOrStorageEntry::Storage(storage_entry)) = era2.next() else {
+                    return Err(Error::msg("Expected storage, got account entry"));
+                };
+                for StorageItem {
+                    storage_index_hash,
+                    value,
+                } in storage_entry.0
+                {
+                    account_trie
+                        .insert(storage_index_hash.as_slice(), &alloy_rlp::encode(value))?;
+                }
+            }
+
+            if account_trie.root_hash()? != account_state.storage_root {
+                return Err(Error::msg("Failed importing account storage trie: storage roots don't match expect value, .era2 import failed"));
+            }
+
+            // Insert contract if available
+            if !bytecode.is_empty() && account_state.code_hash != KECCAK_EMPTY {
+                ensure!(
```

Comment:
nit: With this check being inside `if`, it's possible that we will accept invalid era2 file, for example: bytecode provided with this account is empty but code_hash is not (and serious attack vector if the same code_hash is hash of bytecode of some other account).

Therefore, I think it's better to move this check before `if` statement. And with this check present, we don't need both conditions in the `if` statement (`!bytes.is_empty()` is enough).